### DISCUSSION TITLE
Best way to determine if a query/select has an oder by clause

### DISCUSSION THREAD

**Discussion by olliemath** [sqlalchemy/sqlalchemy#10515] [Usage Questions] [answered]
Hello all, I'm trying to find out the best way to determine if an existing query or select object has an order by clause.

I've found this stack overflow answer: https://stackoverflow.com/questions/67477889/given-a-sqlalchemy-query-object-how-can-i-determine-if-and-how-it-has-been-order - is that the recommended way or is there a better method, e.g. using inspect?

## Some background and motivation

For background I'm working on an existing large suite of tests (1000s) and trying to reduce their flakiness. One simple cause of flake is that running `session.execute(select(SomeModel)).all()` against a postgres instance will _usually_ return objects in the order they were added to the session. But _occasionally_ they come back in a different order 🎲. That means many tests have been written, which seemed to pass, but without an explicit order_by will occasionally flake.

My idea was basically to subclass and add an `oder_by(func.random())` to clauses that don't already have an explicit order during testing (thereby highlighting all the bad tests and generally forcing contributors to deal with this semi-randomness much earlier). But to do that I need to determine the "good" select statements and leave them alone.

**Comment by zzzeek** [comment, marked as answer]
if you are trying to detect unordered-ness in SELECT statements, you probably want to add the random order by to *all* statements; even if a statement has ordering on it, it might not be ordered for every column that's needed.

in SQLAlchemy 1.4 and 2.0 all ORM Query objects are turned into select(), so if you implement a hook like [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) for ORM or [before_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.before_execute) for Core you can intercept all select() objects before they are compiled.  

If you do want to see existing ordering, look for the [_order_by_clauses](https://github.com/sqlalchemy/sqlalchemy/blob/2318d6631de3d57c9ba1a6b627285c0ed46a4a09/lib/sqlalchemy/sql/selectable.py#L3756) tuple, which is semi private.

To add ordering to any of these statements, do `stmt = stmt.order_by(...random...)`.

**Reply by olliemath** [reply]
Thanks for the speedy answer, that's super useful. I had sort of missed the fact that the random order would be a refinement of any existing order.

Adding it to all statements will certainly be easier to implement!

Off to see how many tests I can break...
